auth_fastapi/session.py

```diff
@@ -0,0 +1,47 @@
+import edgedb
+
+from typing import Annotated, Optional, Union
+from fastapi import Cookie, Depends
+
+ClientDep = Annotated[edgedb.AsyncIOClient, Depends(edgedb.create_async_client)]
+
+
+class BaseSession:
+    client: edgedb.AsyncIOClient
+
+    def __init__(self, *, client: edgedb.AsyncIOClient):
+        self.client = client
+
+    async def is_authenticated(self) -> bool:
+        return await self.client.query_required_single(  # type: ignore
+            "select exists ext::auth::ClientTokenIdentity"
+        )
+
+class AuthenticatedSession(BaseSession):
+    auth_token: str
+
+    def __init__(self, *, client: edgedb.AsyncIOClient, auth_token: str):
+        self.auth_token = auth_token
+        self.client = client.with_globals(  # type: ignore
+            {"ext::auth::ClientTokenIdentity": auth_token}
+        )
+
+
+class AnonymousSession(BaseSession):
+    pass
+
+
+Session = Union[AuthenticatedSession, AnonymousSession]
```

Comment:
I think my TypeScripterness is showing through here 😅 . Is this idiomatic?

Reply:
This ADT is probably fine; though the `*FailedResponse` should probably be actual exceptions that are recovered in a separate code path.